### PR TITLE
[React] Settings interface

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Implementing a settings interface for the new UI

### TODO

- [x] Integrate features from https://github.com/inventree/InvenTree/pull/5326
- [x] Implement boolean settings
- [x] Implement number settings
- [x] Implement string settings
- [x] Implement choice settings
- [ ] Implement related model settings
- [x] Global settings
- [x] User settings
- [ ] Plugin settings
- [x] Implement global settings context (zustand, not context)
- [x] Implement user roles context (zustand)
- [ ] split settings list into groups by sub categories 
- [x] use [state manager](https://medium.com/@viraj.vimu/react-context-api-vs-zustand-state-manager-98ca9ac76904) instead of `useContext`
- [x] Change `useInstance` hook to use proper API URLs - not raw strings
- [x] Change form props to use proper API URLs - not raw strings

### CODE REVIEW

File: src/frontend/src/components/nav/MainMenu.tsx [hunk 1]
```diff
@@ -4,46 +4,60 @@ import {
...
       <Menu.Target>
         <UnstyledButton className={classes.layoutHeaderUser}>
           <Group spacing={7}>
-            <Text weight={500} size="sm" sx={{ lineHeight: 1 }} mr={3}>
-              {username ? (
-                username
-              ) : (
-                <Skeleton height={20} width={40} radius={theme.defaultRadius} />
-              )}
-            </Text>
```
Comment (matmair) [inline src/frontend/src/components/nav/MainMenu.tsx:31]: Question: Why is the username removed? How is the user expected to know which user they are currently are using

File: src/frontend/src/components/nav/MainMenu.tsx [hunk 2]
```diff
@@ -4,46 +4,60 @@ import {
...
 import { doClassicLogout } from '../../functions/auth';
 import { InvenTreeStyle } from '../../globalStyle';
-import { useApiState } from '../../states/ApiState';
-import { PlaceholderPill } from '../items/Placeholder';
 
 export function MainMenu() {
   const { classes, theme } = InvenTreeStyle();
-  const [username] = useApiState((state) => [state.user?.name]);
+
+  const user = useContext(UserContext);
```
Comment (matmair) [inline src/frontend/src/components/nav/MainMenu.tsx:22]: Question: Why this switch? Contexts are less efficient with rerenders than using a zustand state. See https://medium.com/@viraj.vimu/react-context-api-vs-zustand-state-manager-98ca9ac76904

File: src/frontend/src/hooks/UseInstance.tsx [hunk 3]
```diff
@@ -1,59 +1,96 @@
...
-  const [instance, setInstance] = useState<any>({});
+export function useInstance({
+  url,
+  pk,
+  params = {},
+  defaultValue = {},
+  fetchOnMount = false,
+  hasPrimaryKey = true
+}: {
+  url: string;
```
Comment (matmair) [inline src/frontend/src/hooks/UseInstance.tsx:30]: Comment: Using strings as URL references will make supporting #1297 very hard

File: src/frontend/src/components/nav/MainMenu.tsx (hunk 1 above)
Comment (SchrodingersGat) [reply src/frontend/src/components/nav/MainMenu.tsx:31]: Oversight on my part, changed that but did not mean to commit

File: src/frontend/src/components/nav/MainMenu.tsx (hunk 2 above)
Comment (SchrodingersGat) [reply src/frontend/src/components/nav/MainMenu.tsx:22]: Thanks, I'll look into that. I was trying to work out how your implementation worked, if that is more efficient then I am all for it.

File: src/frontend/src/hooks/UseInstance.tsx (hunk 3 above)
Comment (SchrodingersGat) [reply src/frontend/src/hooks/UseInstance.tsx:30]: Fair point. I'll make this use your code :)

Comment (SchrodingersGat) [reply src/frontend/src/hooks/UseInstance.tsx:30]: @matmair for reference, a refactor of the URLs - https://github.com/inventree/InvenTree/pull/5686

**Review by matmair — APPROVED** [review]
This is a good start for building out all required management options via the API; Great work

File: src/frontend/src/components/settings/SettingItem.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      },
+      onFormSuccess() {
+        showNotification({
+          title: t`Setting updated`,
+          message: t`${setting?.name} updated successfully`,
+          color: 'green'
+        });
+        settingsState.fetchSettings();
+      }
+    });
```
[the remark this answers is not in the document]
Comment (wolflu05) [reply src/frontend/src/components/settings/SettingItem.tsx:77]: I really like the idea, a modal requires way more clicks which is bad ux

[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/frontend/src/components/settings/SettingItem.tsx:77]: The main drivers for a modal here is that the modal framework is already in place which handles API transactions, error handling, etc

I'd happily review an inline editor solution - potentially the forms components I have already implemented could be reused. The only requirement would be that we add provision for an in-line error message which is displayed when the value is marked as invalid by the server.

**Comment by SchrodingersGat** [comment]
@matmair I am going to merge this in now as I think it is sufficiently advanced that we can leverage it. Thanks for your good points and improvements, I think it is pretty clean now!

Note that this does not address all the points in #5326 - perhaps the task-list over there can be moved to a new issue so we can continue to tick off the other items?

**Pull request merged**